### PR TITLE
Move the View and Controller to ElixirTools

### PR DESCRIPTION
**Pull request by hylkealons**
#### :tophat: Problem
Json API schemas is the standard we have between our applications. Moving it to the ElixirTools folder makes that we can share it, and implement it faster and easier into our applications.

#### :pushpin: Solution
This is the first PR out of 2. This PR introduces the ja_serializer library, the controller and the view. The second PR will contain error handling and the usage of the Plug.

### CODE REVIEW

File: lib/json_api/view.ex [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  def render(serializer, data) do
+    serializer
+    |> JaSerializer.PhoenixView.render(data)
+    |> format()
+  end
+
+  @spec format(map) :: map
+  defp format(map) do
+    for {key, value} <- ensure_map(map), into: %{} do
+      key = convert_key(key)
```
Comment (fartek) [inline lib/json_api/view.ex]: Opinions on variable rebinding? Do we use it often in our codebase? In my opinion it's a bit easier to debug when variable names are explicit, eg. `converted_key = convert_key(key)`. It's easier to keep track and follow the code.

File: lib/json_api/view.ex [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  def render(serializer, data) do
+    serializer
+    |> JaSerializer.PhoenixView.render(data)
+    |> format()
+  end
+
+  @spec format(map) :: map
+  defp format(map) do
+    for {key, value} <- ensure_map(map), into: %{} do
+      converted_key = convert_key(key)
```
Comment (Kitton) [inline lib/json_api/view.ex]: it's inconsistent with the next line now

File: lib/json_api/controller.ex [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    def my_method(conn, phoenix_params, json_params), do: conn
+  end
+  ```
+  """
+
+  @type params :: map
+
+  defmacro __using__(_) do
+    quote do
+      def action(conn, _) do
```
Comment (Kitton) [inline lib/json_api/controller.ex]: why removed?
```elixir
      @spec action(Plug.Conn.t(), any) :: Plug.Conn.t()
```

File: lib/json_api/view.ex [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+      def format_datetime(value) do
+        value
+        |> Timex.Timezone.convert("UTC")
+        |> DateTime.to_iso8601()
+      end
+    end
+  end
+
+  def render(serializer, data) do
```
Comment (Kitton) [inline lib/json_api/view.ex]: probably it's a good time to add missing specs to this file, wdyt @hylkealons ?

File: lib/json_api/view.ex [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+  @spec render(module, map) :: map
+  def render(serializer, data) do
+    serializer
+    |> JaSerializer.PhoenixView.render(data)
+    |> format()
+  end
+
+  @spec format(map) :: map
+  defp format(map) do
```
Comment (fartek) [inline lib/json_api/view.ex]: I would suggest having 2 format functions.
The first one being
```
defp format(map) when is_map(map) do
  for {key, value} <- map, into: %{} do
      {convert_key(key), convert_value(value)}
  end
end
```
and second one
```
defp format(%{__struct__: _} = val) do
  val |> ensure_map |> format
end
```

I think this makes it a bit clearer than having a function parameter named `map` which might or might not be a map. This way will also I think fail with a clearer error message if a parameter of the wrong type is passed.
Thoughts on this?

File: lib/json_api/view.ex [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  @spec convert_key(String.t() | atom) :: String.t()
+  defp convert_key(key), do: ensure_dash(to_string(key))
+
+  @spec ensure_dash(any) :: any
+  defp ensure_dash(value) when is_binary(value), do: String.replace(value, "_", "-")
+  defp ensure_dash(value), do: value
+
+  @spec ensure_map(struct | map) :: map
+  defp ensure_map(%{__struct__: _} = struct), do: Map.from_struct(struct)
+  defp ensure_map(map), do: map
```
Comment (fartek) [inline lib/json_api/view.ex]: why not `when is_map(map)` here, and let it crash for other/invalid types?

File: lib/json_api/view.ex (hunk 5 above)
Comment (hylkealons) [reply lib/json_api/view.ex]: not related to this PR, in which I am just moving the existing logic to Elixir Tools. Feel free to refactor after though or by creating an issue: what you say makes some sense.

File: lib/json_api/view.ex (hunk 6 above)
Comment (hylkealons) [reply lib/json_api/view.ex]: by default it's a collectable

**Pull request merged**